filclient/filclient.go

```diff
@@ -853,221 +857,113 @@ func (fc *FilClient) RetrieveContent(ctx context.Context, miner address.Address,
+	// Submit the retrieval deal proposal to the miner
+	chanid, err := fc.dataTransfer.OpenPullDataChannel(ctx, mpID, proposal, proposal.PayloadCID, shared.AllSelector())
 	if err != nil {
 		return nil, err
 	}
 
-	unsub := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, channelState datatransfer.ChannelState) {
-		if channelState.ChannelID() == chanid {
-			fmt.Printf("[%s] event: %s %d %s %s\n", time.Now().Format("15:04:05"), event.Message, event.Code, datatransfer.Events[event.Code], event.Timestamp.Format("15:04:05"))
-			fmt.Printf("channelstate: %s %s\n", datatransfer.Statuses[channelState.Status()], channelState.Message())
+	// Used to limit prints to the console about received status
+	lastReceivedUpdate := time.Now()
+
+	// The next nonce (incrementing unique ID starting from 0) for the next voucher
+	var nonce uint64 = 0
+
+	// Set up incoming events handler
+
+	// dtRes receives either an error (failure) or nil (success) which is waited
+	// on and handled below before exiting the function
+	dtRes := make(chan error)
+	unsubscribe := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, state datatransfer.ChannelState) {
+		switch event.Code {
+		case datatransfer.NewVoucherResult:
+			switch resType := state.LastVoucherResult().(type) {
+			case *retrievalmarket.DealResponse:
+				switch resType.Status {
+				case retrievalmarket.DealStatusAccepted:
+					log.Info("deal accepted")
+
+				// Respond with a payment voucher when funds are requested
+				case retrievalmarket.DealStatusFundsNeeded:
+					log.Infof("sending payment voucher: %v", resType.PaymentOwed)
+
+					totalPayment = types.BigAdd(totalPayment, resType.PaymentOwed)
+
+					vres, err := fc.pchmgr.CreateVoucher(ctx, pchAddr, paych.SignedVoucher{
+						ChannelAddr: pchAddr,
+						Lane:        pchLane,
+						Nonce:       nonce,
+						Amount:      totalPayment,
+					})
+					if err != nil {
+						dtRes <- err
+					}
+
+					if types.BigCmp(vres.Shortfall, big.NewInt(0)) > 0 {
+						dtRes <- xerrors.Errorf("not enough funds remaining in payment channel (shortfall = %s)", vres.Shortfall)
+					}
+
+					if err := fc.dataTransfer.SendVoucher(ctx, chanid, &retrievalmarket.DealPayment{
+						ID:             proposal.ID,
+						PaymentChannel: pchAddr,
+						PaymentVoucher: vres.Voucher,
+					}); err != nil {
+						dtRes <- xerrors.Errorf("failed to send payment voucher: %w", err)
+					}
+
+					nonce++
+				case retrievalmarket.DealStatusFundsNeededUnseal:
+					dtRes <- xerrors.Errorf("received unexpected payment request for unsealing data")
+				default:
+					log.Debugf("unrecognized voucher: %v", resType)
+				}
+			default:
+				log.Error("unrecognized voucher response type")
+			}
+		case datatransfer.DataReceived:
+			if time.Since(lastReceivedUpdate) >= time.Millisecond*100 {
+				fmt.Printf("received: %v\r", state.Received())
+				lastReceivedUpdate = time.Now()
+			}
+		case datatransfer.FinishTransfer:
```

Comment:
Probably worth having a 'default' case to this outer switch statement , not sure what else could happen here, but i figure it might be worth paying attention to.

Reply:
added